src/core/mesh/qgsmeshtracerenderer.cpp

```diff
@@ -0,0 +1,605 @@
+    fieldWidth++;
+  if ( fieldHeightInDeviceCoordinate % mFieldResolution > 0 )
+    fieldHeight++;
+
+  mFieldSize.setWidth( fieldWidth );
+  mFieldSize.setHeight( fieldHeight );
+
+
+  double mapUnitperFieldPixel;
+  if ( interestZoneExtent.width() > 0 )
+    mapUnitperFieldPixel = interestZoneExtent.width() / fieldWidthInDeviceCoordinate * mFieldResolution;
+  else
+    mapUnitperFieldPixel = 1e-8;
+
+  int fieldRightDevice = mFieldTopLeftInDeviceCoordinates.x() + mFieldSize.width() * mFieldResolution;
+  int fieldBottomDevice = mFieldTopLeftInDeviceCoordinates.y() + mFieldSize.height() * mFieldResolution;
+  QgsPointXY fieldRightBottomMap = deviceMapToPixel.toMapCoordinates( fieldRightDevice, fieldBottomDevice );
+
+  int fieldTopDevice = mFieldTopLeftInDeviceCoordinates.x();
+  int fieldLeftDevice = mFieldTopLeftInDeviceCoordinates.y();
+  QgsPointXY fieldTopLeftMap = deviceMapToPixel.toMapCoordinates( fieldTopDevice, fieldLeftDevice );
+
+  double xc = ( fieldRightBottomMap.x() + fieldTopLeftMap.x() ) / 2;
+  double yc = ( fieldTopLeftMap.y() + fieldRightBottomMap.y() ) / 2;
+
+  mMapToFieldPixel = QgsMapToPixel( mapUnitperFieldPixel,
+                                    xc,
+                                    yc,
+                                    fieldWidth,
+                                    fieldHeight, 0 );
+
+  initField();
+  mValid = true;
+}
+
+void QgsMeshTraceField::addTrace( QgsPointXY startPoint )
+{
+  addTrace( mMapToFieldPixel.transform( startPoint ).toQPointF().toPoint() );
+}
+
+void QgsMeshTraceField::addRandomTraces( int count )
+{
+  for ( int i = 0; i < count; ++i )
+    addRandomTrace();
+}
+
+void QgsMeshTraceField::addRandomTrace()
+{
+  if ( !mValid )
+    return;
+
+  int xRandom =  1 + std::rand() / int( ( RAND_MAX + 1u ) / uint( mFieldSize.width() ) )  ;
+  int yRandom = 1 + std::rand() / int ( ( RAND_MAX + 1u ) / uint( mFieldSize.height() ) ) ;
+
+  addTrace( QPoint( xRandom, yRandom ) );
+}
+
+void QgsMeshTraceField::addGriddedTraces( int pixelSpace )
+{
+  int fieldSpace = pixelSpace / mFieldResolution;
+  int i = 0;
+  while ( i < size().width() )
+  {
+    int j = 0;
+    while ( j < size().height() )
+    {
+      addTrace( QPoint( i, j ) );
+      j += fieldSpace;
+    }
+    i += fieldSpace;
+  }
+}
+
+void QgsMeshTraceField::addTracesFromBorder( int pixelSpace )
+{
+  int fieldSpace = pixelSpace / mFieldResolution;
+  int i = 0;
+  int j = 0;
+  while ( j < size().height() )
+  {
+    addTrace( QPoint( i, j ) );
+    j += fieldSpace;
+  }
+  i = size().width() - 1;
+  j = 0;
+  while ( j < size().height() )
+  {
+    addTrace( QPoint( i, j ) );
+    j += fieldSpace;
+  }
+  i = 0;
+  j = 0;
+  while ( i < size().width() )
+  {
+    addTrace( QPoint( i, j ) );
+    i += fieldSpace;
+  }
+  i = 0;
+  j = size().height() - 1;
+  while ( i < size().width() )
+  {
+    addTrace( QPoint( i, j ) );
+    i += fieldSpace;
+  }
+
+}
+
+void QgsMeshTraceField::addTrace( QPoint startPixel )
```

Comment:
can you write docs/description how this method works (maybe (hand-drawn) picture?) it is hard to undrestand the logic without some description. thanks

Reply:
The algorithm will may be change a bit, especially to reduce aliasing in the rendering. But promise, I will make a clear(?) doc one day.